File: TrumpetPython/services/items_service.py
from typing import List, Optional, Any, Dict
from sqlmodel import Session, select, func, or_, and_
from sqlalchemy.orm import selectinload
from starlette.concurrency import run_in_threadpool
from functools import lru_cache
from models.item import Item
from models.metadata_value import MetadataValue
from models.collection import Collection

class ItemsService:
    _cache = {}

    def __init__(self, session: Session):
        self.session = session

    async def get_fields(self) -> List[str]:
        if "fields" in self._cache:
            return self._cache["fields"]
        statement = select(MetadataValue.Field).distinct()
        results = await run_in_threadpool(lambda: self.session.exec(statement).all())
        self._cache["fields"] = list(results)
        return self._cache["fields"]

    async def get_path_values(self) -> List[Optional[str]]:
        statement = select(MetadataValue.Value).where(
            and_(MetadataValue.Field == "dc.musicsubpath", MetadataValue.Language == "en")
        ).distinct()
        results = await run_in_threadpool(lambda: self.session.exec(statement).all())
        return list(results)

    async def get_path_counts(self) -> List[Dict[str, Any]]:
        if "path_counts" in self._cache:
            return self._cache["path_counts"]
        statement = (
            select(MetadataValue.Value, func.count(MetadataValue.Id))
            .where(and_(MetadataValue.Field == "dc.musicsubpath", MetadataValue.Language == "en"))
            .group_by(MetadataValue.Value)
        )
        results = await run_in_threadpool(lambda: self.session.exec(statement).all())
        self._cache["path_counts"] = [{"value": row[0], "count": row[1]} for row in results]
        return self._cache["path_counts"]

    async def search_all_metadata(self, value: str) -> List[Dict[str, Any]]:
        statement = (
            select(MetadataValue.Field, func.count(MetadataValue.Id))
            .where(MetadataValue.Value.contains(value))
            .group_by(MetadataValue.Field)
        )
        results = await run_in_threadpool(lambda: self.session.exec(statement).all())
        return [{"field": row[0], "count": row[1]} for row in results]

    async def get_items(
        self, 
        path: Optional[str] = None, 
        search: Optional[str] = None, 
        community_id: Optional[str] = None, 
        collection_id: Optional[str] = None, 
        page: int = 1, 
        page_size: int = 10
    ) -> List[Item]:
        # Using selectinload to eagerly load metadata and bitstreams (similar to .Include in EF Core)
        statement = select(Item).options(
            selectinload(Item.metadata_entries),
            selectinload(Item.bitstreams)
        )

        if path:
            search_values = []
            if path == "ArtMusic":
                search_values = ["Art music", "Μουσική του άστεως", "Μουσική του Άστεως", "Art"]
            elif path == "UrbanPopular":
                search_values = ["Urban popular music", "Αστικολαϊκή μουσική", "Urban"]
            elif path == "RuralMusic":
                search_values = ["Rural music", "Μουσική της υπαίθρου", "Rural"]
            elif path == "SacredMusic":
                search_values = ["Sacred music", "Εκκλησιαστική μουσική", "Sacred"]
            
            if search_values:
                # Filter items that have metadata with matching values
                sub_query = select(MetadataValue.ItemId).where(
                    and_(
                        MetadataValue.Field == "dc.musicsubpath",
                        or_(*[MetadataValue.Value.contains(sv) for sv in search_values])
                    )
                )
                statement = statement.where(Item.Id.in_(sub_query))

        if community_id:
            # Filter by community (via collection)
            coll_sub_query = select(Collection.Id).where(Collection.ParentCommunityId == community_id)
            statement = statement.where(Item.CollectionId.in_(coll_sub_query))

        if collection_id:
            statement = statement.where(Item.CollectionId == collection_id)

        if search:
            # Search in Name or any Metadata value
            meta_sub_query = select(MetadataValue.ItemId).where(MetadataValue.Value.contains(search))
            statement = statement.where(
                or_(
                    Item.Name.contains(search),
                    Item.Id.in_(meta_sub_query)
                )
            )

        # Pagination
        statement = statement.order_by(Item.Name).offset((page - 1) * page_size).limit(page_size)
        
        results = await run_in_threadpool(lambda: self.session.exec(statement).all())
        return list(results)

    async def get_item(self, id: str) -> Optional[Item]:
        statement = select(Item).where(Item.Id == id).options(
            selectinload(Item.metadata_entries),
            selectinload(Item.bitstreams)
        )
        result = await run_in_threadpool(lambda: self.session.exec(statement).first())
        return result

File: data/backend/src/Services/ItemsService.php
<?php

declare(strict_types=1);

namespace Trumpet\Services;

use PDO;

/**
 * ItemsService
 *
 * Mirrors items_service.py — all queries translated from SQLModel/SQLAlchemy to PDO.
 * Path filtering uses the same string-match lists as the Python version.
 */
class ItemsService
{
    public function __construct(private readonly PDO $pdo)
    {
    }

    // ── Fields ──────────────────────────────────────────────────────────────

    public function getFields(): array
    {
        $stmt = $this->pdo->query('SELECT DISTINCT `Field` FROM MetadataValues ORDER BY `Field`');
        return $stmt->fetchAll(PDO::FETCH_COLUMN);
    }

    // ── Path values ─────────────────────────────────────────────────────────

    public function getPathValues(): array
    {
        $stmt = $this->pdo->prepare(
            "SELECT DISTINCT `Value` FROM MetadataValues
             WHERE `Field` = 'dc.musicsubpath' AND Language = 'en'
             ORDER BY `Value`"
        );
        $stmt->execute();
        return $stmt->fetchAll(PDO::FETCH_COLUMN);
    }

    // ── Path counts ─────────────────────────────────────────────────────────

    public function getPathCounts(): array
    {
        $stmt = $this->pdo->prepare(
            "SELECT `Value` AS `value`, COUNT(Id) AS `count`
             FROM MetadataValues
             WHERE `Field` = 'dc.musicsubpath' AND Language = 'en'
             GROUP BY `Value`
             ORDER BY `count` DESC"
        );
        $stmt->execute();
        return $stmt->fetchAll();
    }

    // ── Search all metadata ─────────────────────────────────────────────────

    public function searchAllMetadata(string $value): array
    {
        $stmt = $this->pdo->prepare(
            "SELECT `Field` AS `field`, COUNT(Id) AS `count`
             FROM MetadataValues
             WHERE `Value` LIKE :val
             GROUP BY `Field`
             ORDER BY `count` DESC"
        );
        $stmt->execute([':val' => '%' . $value . '%']);
        return $stmt->fetchAll();
    }

    // ── Get items (paginated, filtered) ─────────────────────────────────────

    public function getItems(
        ?string $path         = null,
        ?string $search       = null,
        ?string $communityId  = null,
        ?string $collectionId = null,
        int     $page         = 1,
        int     $pageSize     = 10
    ): array {
        $where  = [];
        $params = [];

        // Path filter — mirrors items_service.py logic
        if ($path !== null) {
            $searchValues = match ($path) {
                'ArtMusic'     => ['Art music', 'Μουσική του άστεως', 'Μουσική του Άστεως', 'Art'],
                'UrbanPopular' => ['Urban popular music', 'Αστικολαϊκή μουσική', 'Urban'],
                'RuralMusic'   => ['Rural music', 'Μουσική της υπαίθρου', 'Rural'],
                'SacredMusic'  => ['Sacred music', 'Εκκλησιαστική μουσική', 'Sacred'],
                default        => [],
            };

            if (!empty($searchValues)) {
                $likes = [];
                foreach ($searchValues as $i => $sv) {
                    $key    = ":path_val_{$i}";
                    $likes[]= "mv_path.Value LIKE {$key}";
                    $params[$key] = '%' . $sv . '%';
                }
                $where[] = "i.Id IN (
                    SELECT mv_path.ItemId FROM MetadataValues mv_path
                    WHERE mv_path.Field = 'dc.musicsubpath'
                    AND (" . implode(' OR ', $likes) . ")
                )";
            }
        }

        // Community filter (via collection)
        if ($communityId !== null) {
            $where[]              = 'c.ParentCommunityId = :community_id';
            $params[':community_id'] = $communityId;
        }

        // Collection filter
        if ($collectionId !== null) {
            $where[]               = 'i.CollectionId = :collection_id';
            $params[':collection_id'] = $collectionId;
        }

        // Full-text search
        if ($search !== null) {
            $where[]         = "(i.Name LIKE :search OR i.Id IN (
                SELECT mv_s.ItemId FROM MetadataValues mv_s WHERE mv_s.Value LIKE :search2
            ))";
            $params[':search']  = '%' . $search . '%';
            $params[':search2'] = '%' . $search . '%';
        }

        $whereClause = $where ? ('WHERE ' . implode(' AND ', $where)) : '';

        $offset = ($page - 1) * $pageSize;
        $sql    = "
            SELECT i.Id, i.Name, i.Handle, i.LastModified, i.Withdrawn, i.Archived, i.CollectionId
            FROM Items i
            LEFT JOIN Collections c ON i.CollectionId = c.Id
            {$whereClause}
            ORDER BY i.Name
            LIMIT :page_size OFFSET :offset
        ";

        $stmt = $this->pdo->prepare($sql);
        foreach ($params as $k => $v) {
            $stmt->bindValue($k, $v);
        }
        $stmt->bindValue(':page_size', $pageSize, PDO::PARAM_INT);
        $stmt->bindValue(':offset',    $offset,   PDO::PARAM_INT);
        $stmt->execute();

        $items = $stmt->fetchAll();
        return $this->hydratePublic($items);
    }

    // ── Get single item ─────────────────────────────────────────────────────

    public function getItem(string $id): ?array
    {
        $stmt = $this->pdo->prepare(
            'SELECT Id, Name, Handle, LastModified, Withdrawn, Archived, CollectionId
             FROM Items WHERE Id = :id'
        );
        $stmt->execute([':id' => $id]);
        $item = $stmt->fetch();
        if (!$item) {
            return null;
        }
        $items = $this->hydratePublic([$item]);
        return $items[0] ?? null;
    }

    // ── Hydration helpers ───────────────────────────────────────────────────

    /**
     * Attach metadata and bitstreams to a list of raw item rows in bulk.
     * Uses 2 extra queries (IN clause) rather than N+1.
     * Public so CommunityItemsService, TaggingService, RecommendationService can reuse it.
     */
    public function hydratePublic(array $items): array
    {
        if (empty($items)) {
            return [];
        }

        $ids = array_column($items, 'Id');
        $placeholders = implode(',', array_fill(0, count($ids), '?'));

        // Metadata
        $metaStmt = $this->pdo->prepare(
            "SELECT Id, ItemId, `Field`, `Value`, Language
             FROM MetadataValues WHERE ItemId IN ({$placeholders}) ORDER BY Id"
        );
        $metaStmt->execute($ids);
        $metaRows = $metaStmt->fetchAll();

        $metaByItem = [];
        foreach ($metaRows as $row) {
            $metaByItem[$row['ItemId']][] = [
                'id'       => $row['Id'],
                'itemId'   => $row['ItemId'],
                'field'    => $row['Field'],
                'value'    => $row['Value'],
                'language' => $row['Language'],
            ];
        }

        // Bitstreams
        $bsStmt = $this->pdo->prepare(
            "SELECT Id, ItemId, Name, MimeType, SizeBytes, LocalFilePath
             FROM Bitstreams WHERE ItemId IN ({$placeholders}) ORDER BY Name"
        );
        $bsStmt->execute($ids);
        $bsRows = $bsStmt->fetchAll();

        $bsByItem = [];
        foreach ($bsRows as $row) {
            $bsByItem[$row['ItemId']][] = [
                'id'            => $row['Id'],
                'itemId'        => $row['ItemId'],
                'name'          => $row['Name'],
                'mimeType'      => $row['MimeType'],
                'sizeBytes'     => (int) $row['SizeBytes'],
                'localFilePath' => $row['LocalFilePath'],
            ];
        }

        // Compose final output
        $result = [];
        foreach ($items as $item) {
            $id = $item['Id'];
            $result[] = [
                'id'           => $id,
                'name'         => $item['Name'],
                'handle'       => $item['Handle'],
                'lastModified' => $item['LastModified'],
                'withdrawn'    => (bool) $item['Withdrawn'],
                'archived'     => (bool) $item['Archived'],
                'collectionId' => $item['CollectionId'],
                'metadata'     => $metaByItem[$id] ?? [],
                'bitstreams'   => $bsByItem[$id]   ?? [],
            ];
        }

        return $result;
    }
}
